From 97c3a03fb13c9271c52877e2f3805a06ab298db0 Mon Sep 17 00:00:00 2001
From: martinbaker <ax87438@martinb.com>
Date: Sun, 18 Sep 2016 16:30:51 +0100
Subject: [PATCH 4/4] fix regression bug

---
 src/algebra/computation.spad | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/algebra/computation.spad b/src/algebra/computation.spad
index 591ca94..fdf75ea 100644
--- a/src/algebra/computation.spad
+++ b/src/algebra/computation.spad
@@ -2230,7 +2230,7 @@ SKICombinators(UT) : Exports == Implementation where
       s : OutputForm := empty()$OutputForm
       if n case lf then
           -- leaf node so print I, K or S
-          string(n.lf)::OutputForm
+          s := (n.lf)::OutputForm
       if n case nd then
           -- binary node which has two subnodes c1 and c2
           if atom?(n.nd.c2) then
-- 
2.1.4

